File: economy/moneyLog.js
/* ---------------- economy/moneyLog: balance-change watcher ----------------
   Pavlov's economy mod keeps each player's money in its own ModSave text file
   (`<PlayerName>.txt`, containing a bare integer). Nothing announces a change - the
   game just rewrites the file. This watches them and reports every movement:

     +400 to player1
     -345 to player2

   Design notes:

   SEEDS SILENTLY. The first scan records balances WITHOUT reporting them, otherwise
   every restart would announce one line per player as if everyone had just been paid.

   USES mtime. Re-reading every ledger on each tick is pointless disk work when almost
   none have changed. Files whose mtime is unchanged are skipped entirely, so the poll
   stays cheap with hundreds of players.

   COLLAPSES BULK CHANGES. A wipe or a mass payout would otherwise emit one line per
   player - hundreds of messages, rate-limited into uselessness, burying anything real.
   Past a threshold it reports a summary instead.

   The scan/diff/format helpers are pure and exported, so the interesting logic tests
   without a filesystem or a webhook. */
module.exports = function(ctx) {
  const {
  fs, path, logger, getModsavePath, MODSAVE_SYNC_SKIP, postMoneyLog,
  } = ctx;

  /* Beyond this many changes in one tick, report a count rather than a list. A real
     round produces a handful of movements; hundreds means a wipe or a bulk operation. */
  const BULK_CHANGE_LIMIT = 12;

  /** Is this directory entry one of the per-player ledger files? */
  function isLedgerFile(name, isFile) {
    if (!isFile) return false;                                  // FactionRoles, RCON Plus, ...
    const lower = String(name).toLowerCase();
    if (!lower.endsWith(".txt")) return false;
    if (lower === "banlist.txt") return false;                  // ban-message file, not a ledger
    if (MODSAVE_SYNC_SKIP && MODSAVE_SYNC_SKIP.test(name)) return false;   // menu-access files
    return true;
  }

  /**
   * Read every ledger into a Map of player -> { balance, mtimeMs }.
   * `previous` lets unchanged files be skipped: same mtime means same contents, so the
   * cached balance is reused rather than re-parsed.
   */
  function scanLedgers(previous = new Map()) {
    const base = getModsavePath();
    if (!base) return null;                                     // economy not configured
    let entries;
    try { entries = fs.readdirSync(base, { withFileTypes: true }); }
    catch (e) { logger.warn("MoneyLog", `cannot read ${base}: ${e.code || e.message}`); return null; }

    const out = new Map();
    for (const ent of entries) {
      if (!isLedgerFile(ent.name, ent.isFile())) continue;
      const player = ent.name.slice(0, -4);                     // strip ".txt"
      const fp = path.join(base, ent.name);

      let mtimeMs;
      try { mtimeMs = fs.statSync(fp).mtimeMs; }
      catch { continue; }                                       // vanished mid-scan - next tick will see it

      const cached = previous.get(player);
      if (cached && cached.mtimeMs === mtimeMs) { out.set(player, cached); continue; }

      let balance = NaN;
      try { balance = parseInt(fs.readFileSync(fp, "utf8").trim(), 10); }
      catch { /* handled below, same as an unparseable read */ }

      /* An unparseable ledger - a mid-write truncation, a corrupt file - must not be read
         as 0, which would report a huge fabricated DEBIT for a player whose money is fine.
         It must also not simply drop the player, because losing their last known balance
         makes the next valid read look like a brand-new account and reports the whole
         amount as a fabricated CREDIT. Carry the last good entry forward with its OLD
         mtime, so the next tick re-reads and the comparison stays anchored to real data. */
      if (!Number.isFinite(balance)) {
        if (cached) out.set(player, cached);
        continue;
      }

      out.set(player, { balance, mtimeMs });
    }
    return out;
  }

  /**
   * Compare two scans. Returns [{ player, before, after, delta }], biggest movement
   * first. A player appearing for the first time counts as a credit from zero; one
   * disappearing is NOT reported, because a deleted ledger is an admin action
   * (/wipe, a manual clear) rather than a player losing money.
   */
  function diffScans(before, after) {
    const changes = [];
    for (const [player, now] of after) {
      const was = before.get(player);
      const previous = was ? was.balance : 0;
      if (was && was.balance === now.balance) continue;
      if (!was && now.balance === 0) continue;                  // new empty ledger is not a movement
      changes.push({ player, before: previous, after: now.balance, delta: now.balance - previous });
    }
    changes.sort((a, b) => Math.abs(b.delta) - Math.abs(a.delta));
    return changes;
  }

  /** "+400 to player1" / "-345 to player2", one per line. */
  function formatChanges(changes) {
    return changes
      .map(c => `${c.delta > 0 ? "+" : "-"}${Math.abs(c.delta).toLocaleString("en-US")} to ${c.player}`)
      .join("\n");
  }

  // ---- the watcher ----
  let cache = new Map();
  let seeded = false;

  /** One pass. Returns the changes it reported, for tests and diagnostics. */
  function tick() {
    const scan = scanLedgers(cache);
    if (!scan) return [];

    if (!seeded) {
      cache = scan;
      seeded = true;
      logger.info("MoneyLog", `watching ${scan.size} player ledger(s) - baseline recorded, not reported`);
      return [];
    }

    const changes = diffScans(cache, scan);
    cache = scan;
    if (!changes.length) return [];

    if (changes.length > BULK_CHANGE_LIMIT) {
      /* A wipe or bulk payout. One line per player would be hundreds of messages, so
         report the shape of it and leave the detail to the logs. */
      const up = changes.filter(c => c.delta > 0).length;
      const net = changes.reduce((s, c) => s + c.delta, 0);
      logger.warn("MoneyLog", `${changes.length} balances changed at once - reporting a summary`);
      postMoneyLog(`${changes.length} balances changed at once (${up} up, ${changes.length - up} down, net ${net >= 0 ? "+" : "-"}${Math.abs(net).toLocaleString("en-US")})`);
      return changes;
    }

    postMoneyLog(formatChanges(changes));
    return changes;
  }

  return { tick, scanLedgers, diffScans, formatChanges, isLedgerFile, BULK_CHANGE_LIMIT,
           _state: () => ({ seeded, tracked: cache.size }) };
};

File: dotnet/src/PavlovBot.Host/Economy/MoneyLog.cs
using Microsoft.Extensions.Logging;
using PavlovBot.Core.Economy;
using PavlovBot.Host.Discord;

namespace PavlovBot.Host.Economy;

/// <summary>
/// Watching the game's ledger files and reporting every balance change.
/// </summary>
/// <remarks>
/// The game writes one plain-text file per player containing their balance. There is no
/// event, no hook, and no API - so the only way to know a payout happened is to notice the
/// number changed. This polls, diffs against a cache, and posts the deltas.
///
/// THE BUG THIS SHAPE EXISTS TO PREVENT: an unreadable or unparseable ledger must carry the
/// CACHED ENTRY FORWARD, not drop it. Dropping it makes the player look new on the next
/// successful read, so their whole balance is reported as a credit - "+5,500" for a player
/// who was paid 500. A fabricated transaction in a money log is worse than a missed one,
/// because somebody will act on it.
///
/// Modification times are used to skip unchanged files, which is what keeps the scan cheap
/// on a server with hundreds of ledgers.
/// </remarks>
public sealed class MoneyLog
{
    private sealed record CachedBalance(long Balance, DateTimeOffset ModifiedAt);

    private readonly Dictionary<string, CachedBalance> _cache = new(StringComparer.OrdinalIgnoreCase);
    private readonly string? _directory;
    private readonly FeedWebhooks _feeds;
    private readonly ILogger<MoneyLog> _logger;
    private readonly TimeProvider _time;
    private bool _primed;

    public MoneyLog(string? ledgerDirectory, FeedWebhooks feeds, ILogger<MoneyLog> logger, TimeProvider? time = null)
    {
        _directory = ledgerDirectory;
        _feeds = feeds;
        _logger = logger;
        _time = time ?? TimeProvider.System;
    }

    public bool Enabled => !string.IsNullOrWhiteSpace(_directory) && Directory.Exists(_directory);

    /// <summary>Poll once. Returns the changes found, which is what the tests assert on.</summary>
    public async Task<IReadOnlyList<(string Player, long Delta)>> TickAsync(CancellationToken ct = default)
    {
        if (!Enabled) return [];

        var changes = new List<(string Player, long Delta)>();

        foreach (var file in EnumerateLedgers())
        {
            ct.ThrowIfCancellationRequested();

            var player = Path.GetFileNameWithoutExtension(file.Name);
            var cached = _cache.GetValueOrDefault(player);

            // Unchanged since the last look: skip without opening it.
            if (cached is not null && cached.ModifiedAt == file.LastWriteTimeUtc) continue;

            if (!TryReadBalance(file.FullName, out var balance))
            {
                /* THE CRITICAL BRANCH. Carry the cached entry forward WITH ITS OLD mtime, so
                   the next tick retries and - crucially - the player is not treated as new. */
                if (cached is not null) _cache[player] = cached;
                continue;
            }

            _cache[player] = new CachedBalance(balance, file.LastWriteTimeUtc);

            if (cached is null) continue;                    // first sighting is a baseline, not a change
            if (balance == cached.Balance) continue;         // touched but unchanged

            changes.Add((player, balance - cached.Balance));
        }

        if (!_primed)
        {
            /* The first tick builds the baseline. Reporting on it would announce every
               player's entire balance as a credit the moment the bot starts. */
            _primed = true;
            _logger.LogInformation("Money log primed with {Count} ledger(s)", _cache.Count);
            return [];
        }

        if (changes.Count > 0)
            await _feeds.PostMoneyAsync(changes, _time.GetUtcNow(), ct).ConfigureAwait(false);

        return changes;
    }

    private IEnumerable<FileInfo> EnumerateLedgers()
    {
        try
        {
            return new DirectoryInfo(_directory!).EnumerateFiles("*.txt");
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            _logger.LogDebug("Could not list {Directory}: {Message}", _directory, ex.Message);
            return [];
        }
    }

    private bool TryReadBalance(string path, out long balance)
    {
        balance = 0;
        try
        {
            var text = File.ReadAllText(path).Trim();
            // The file is a bare number. Anything else is a partial write or a different
            // file that happened to land here, and either way it is not a balance.
            return long.TryParse(text, System.Globalization.NumberStyles.Integer,
                System.Globalization.CultureInfo.InvariantCulture, out balance);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return false;
        }
    }

    /// <summary>The last known balance, without touching the disk.</summary>
    public long? Cached(string player) => _cache.TryGetValue(player, out var entry) ? entry.Balance : null;

    public int Tracked => _cache.Count;
}

/// <summary>
/// Balances, read from and written to the game's own <c>modsave/&lt;player&gt;.txt</c> files.
/// </summary>
/// <remarks>
/// WRITING WAS TURNED OFF ENTIRELY AND IS NOW BACK UNDER ONE RULE: the bot writes a player's
/// ledger only while that player is OFFLINE.
///
/// The reason it was off is not theoretical. The game holds balances in memory and writes
/// these files itself, so a number the bot wrote while somebody was playing got clobbered on
/// the next in-game save - and a read-modify-write landing across an in-game purchase lost
/// money outright. That is what "balances out of step with what the servers held" meant.
///
/// An offline player has no in-memory balance to overwrite, which removes the race rather
/// than narrowing it. The gate lives HERE, at the single write, rather than at each caller,
/// so payroll, <c>/givecaps</c> and <c>/adjustcaps</c> cannot each forget it differently.
///
/// FOUR REFUSALS, ALL FAIL-CLOSED:
///   - the roster cannot be trusted (RCON down): we do not know who is online, so we do not
///     write. "Cannot tell" is not "they are offline".
///   - the player IS online.
///   - the player has NO ledger: the bot never CREATES one. A file the game did not make is
///     a player the game has never banked, and inventing one is how a phantom balance
///     appears for a mistyped name.
///   - the directory does not exist: a wrong MODSAVE_PATH must not silently start a parallel
///     set of ledgers nobody reads.
///
/// The write itself is temp-file-plus-rename, because the game READS these files too and
/// <see cref="MoneyLog.TryReadBalance"/> already has to cope with half-written ones.
/// </remarks>
public sealed class LedgerFileStore(string? directory, PavlovBot.Host.Rcon.IOnlineRoster? online = null) : IBalanceStore
{
    private string? PathFor(string playerId)
    {
        if (string.IsNullOrWhiteSpace(directory)) return null;

        /* Filenames must match what the GAME writes, and names may contain spaces
           ("Butter Life.txt"). Only path separators and control characters are stripped -
           removing spaces would send wages to a phantom "ButterLife.txt" nobody reads. */
        var safe = new string(playerId.Where(c => !char.IsControl(c) && c != '/' && c != '\\').ToArray()).Trim();
        return safe.Length == 0 ? null : Path.Combine(directory, $"{safe}.txt");
    }

    public long? Read(string playerId)
    {
        var path = PathFor(playerId);
        if (path is null || !File.Exists(path)) return null;

        try
        {
            return long.TryParse(File.ReadAllText(path).Trim(),
                System.Globalization.NumberStyles.Integer,
                System.Globalization.CultureInfo.InvariantCulture, out var value) ? value : null;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            // Null means UNKNOWN, and the ledger treats it as zero at the point of use.
            // Returning 0 here would let a transient read error wipe somebody's balance.
            return null;
        }
    }

    /// <summary>Whether this player's ledger may be written right now. See the class remarks.</summary>
    public bool CanWrite(string playerId) => Refusal(playerId) is null;

    /// <summary>
    /// Why a write would be refused, in words a command can show somebody.
    /// </summary>
    /// <remarks>
    /// Separate from <see cref="Write"/> so a command can explain the refusal BEFORE
    /// attempting it. "Not applied" with no reason is what sent people to read the source.
    /// </remarks>
    public string? Refusal(string playerId)
    {
        if (PathFor(playerId) is not { } path) return "there is no ledger directory configured";
        if (!Directory.Exists(directory)) return $"the ledger directory `{directory}` does not exist";

        /* No roster at all means this is a context with no game servers - tests, tooling.
           Refuse rather than assume, for the same reason as the untrusted case below. */
        if (online is null) return "the bot cannot see who is online, so it will not write a ledger";

        if (!online.IsTrustworthy)
            return "no server has a fresh player list, so the bot cannot tell whether they are online";

        if (online.Online.Contains(playerId.Trim(), StringComparer.OrdinalIgnoreCase))
        {
            return "they are currently in game. The server holds their balance in memory and would " +
                   "overwrite anything written now - this applies once they log off";
        }

        if (!File.Exists(path))
            return "they have no ledger file yet. The bot never creates one - the game does that when they first bank";

        return null;
    }

    /// <summary>
    /// Persist a balance, if and only if the player is offline.
    /// </summary>
    /// <returns>False when refused or when the write failed. The balance is unchanged either way.</returns>
    public bool Write(string playerId, long balance)
    {
        if (Refusal(playerId) is not null) return false;
        if (PathFor(playerId) is not { } path) return false;

        try
        {
            /* TEMP THEN RENAME. The game reads these files, and a reader that catches one
               mid-write sees a truncated number - MoneyLog.TryReadBalance already has to
               handle exactly that. A rename is atomic within a directory, so a reader sees
               either the old value or the new one and never half of either. */
            var temporary = path + ".bot.tmp";
            File.WriteAllText(temporary, balance.ToString(System.Globalization.CultureInfo.InvariantCulture));
            File.Move(temporary, path, overwrite: true);
            return true;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return false;
        }
    }
}
